File: client-ecomm02/src/app/prodotto.service.ts
import { TokenDto } from './dto/token-dto';
import { ListaProdottiDto as ListaProdottiDto } from './dto/lista-prodotti-dto';
import { TokenService } from './token.service';
import { Injectable } from '@angular/core';
import { ProdottoDto } from './dto/prodotto-dto';
import { RicercaDto } from './dto/ricerca-dto';
import { HttpClient } from '@angular/common/http';
import { Observable, Subscription } from 'rxjs';
import { Prodotto } from './model/prodotto';

@Injectable({
  providedIn: 'root'
})
export class ProdottoService {
  urlPath = 'http://localhost:8080';
  prodottoForm: Prodotto = new Prodotto(); // memorizza l'istanza da trattare
  prodottoDto: ProdottoDto = new ProdottoDto();
  listaProdotti: Prodotto[] = []; /* il contenitore che renderizza la tabella contente le risposte dal server*/
  listaDto: ListaProdottiDto = new ListaProdottiDto();
  constructor(private http: HttpClient, private srvToken: TokenService,) { }
  /* Passa al server il RicercaDto contenente la stringa da cercare posiziona i dati
  nella Lista aposita, restituisce una istanza di DtoRicerca per resettare il campo
  nel Template*/
  cerca(ricerca: RicercaDto) {
    let o: Observable<ListaProdottiDto> = this.http.post<ListaProdottiDto>(
      this.urlPath + "/prodotti-find/",
      ricerca
    );
    o.subscribe((risp) => {
      this.listaProdotti= risp.listaProdotti;
    });
  }

  lista() {
    let tik: TokenDto = new TokenDto(this.srvToken.token);
    let o: Observable<ListaProdottiDto> = this.http.post<ListaProdottiDto>(
      this.urlPath + "/lista-prodotti",
      tik
    );
    o.subscribe((risp) => {
      this.listaProdotti = risp.listaProdotti;
    });
  }
  /* in Base alla stato del componente che riceve come parametro setta url per la
  richiesta e invia al server l'istanza da trattare. restituisce la stringa
   per ripristinare lo stato iniziale del component*/
  conferma(state: string) {
    let urlEnd: string;
    this.prodottoDto.prodotto = this.prodottoForm;
    this.prodottoDto.token = this.srvToken.token;
    switch (state) {
      case 'modifica': {
        urlEnd = '/prodotti-update';
        break;
      }
      case 'elimina': {
        urlEnd = '/prodotti-delete';
        break;
      }
      case 'aggiungi': {
        urlEnd = '/prodotti-add';
        break;
      }
    }
    console.log(this.prodottoDto);
    const oss: Observable<ProdottoDto> = this.http.post<ProdottoDto>(this.urlPath + urlEnd, this.prodottoDto);
    const sub: Subscription = oss.subscribe(risp => {
      this.lista();
      this.srvToken.token=risp.token;
      console.log(risp); });
    this.prodottoForm = new Prodotto();
    return 'ricerca';
  }
}

File: server/src/main/java/com/ai/ecom02/service/ProdottoService.java

package com.ai.ecom02.service;


import com.ai.ecom02.dto.ListaProdottoDto;
import com.ai.ecom02.dto.ProdottoDto;
import com.ai.ecom02.dto.RicercaDto;
import com.ai.ecom02.model.Prodotto;
import com.ai.ecom02.model.Token;
import java.util.List;


public interface ProdottoService {
    Prodotto add(Prodotto prodotto);

    void delete(Prodotto prodotto);

    Prodotto update(Prodotto prodotto);

    List<Prodotto> getAll();
    
    Prodotto findById(Prodotto prodotto);
    
     Prodotto findByDescrizione(Prodotto prodotto);

    List<Prodotto> findByDescrizioneLike(Prodotto prodotto);
    
    Prodotto findByCodice(Prodotto prodotto);
    
    List<Prodotto> findByCodiceLike(Prodotto prodotto);
    
    List<Prodotto> findByPrezzoLike(Prodotto prodotto);
    
    List<Prodotto> findByCodiceLikeOrDescrizioneLike(RicercaDto ricerca);
    
    public ListaProdottoDto creaListaProdottoDto(List<Prodotto> listaProdotto, Token t);
    
    public ProdottoDto creaListaProdottoDto(Prodotto prodotto, Token t);
    
    
}

